### PR TITLE
HIVE-28131: Iceberg: Add support for Replace Branch.

### PR DESCRIPTION
**Pull request by ayushtkn**
### What changes were proposed in this pull request?

Add support for replace iceberg branch

### Why are the changes needed?

To allow replacing Iceberg branches

### Does this PR introduce _any_ user-facing change?

Yes, Allows replacing iceberg branches

### Is the change a dependency upgrade?

No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by zhangbutao** [comment]
Do we have any plans to support `create or replace table` syntax?
https://iceberg.apache.org/docs/nightly/spark-ddl/#replace-table-as-select
I think for hive table, we can reuse `overwrite ` to emulate `replace `syntax, and for iceberg table, we can just use iceberg api to do this.  We can support iceberg table first.

File: parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g [hunk 1]
```diff
@@ -513,6 +514,13 @@ alterStatementSuffixRenameBranch
     -> ^(TOK_ALTERTABLE_RENAME_BRANCH $sourceBranch $targetBranch)
     ;
 
+alterStatementSuffixReplaceBranch
+@init { gParent.pushMsg("alter table replace branch", state); }
+@after { gParent.popMsg(state); }
+    : KW_REPLACE KW_BRANCH sourceBranch=Identifier KW_WITH KW_SNAPSHOT? (targetBranch=(Identifier | Number))
```
Comment (zhangbutao) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:520]: https://iceberg.apache.org/docs/nightly/spark-ddl/#alter-table-replace-branch Spark-iceberg `replace branch` syntax has more properties, e.g.:
```
-- REPLACE historical-tag to reference snapshot 4567 and update the retention to 60 days.
ALTER TABLE prod.db.sample REPLACE TAG `historical-tag` AS OF VERSION 4567 RETAIN 60 DAYS
```

Any thought we need also to support these?

File: iceberg/iceberg-handler/src/test/queries/positive/replace_iceberg_branch.q [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+set hive.explain.user=false;
+set hive.fetch.task.conversion=more;
+
+create external table ice01(id int) stored by iceberg stored as orc tblproperties ('format-version'='2');
+
+insert into ice01 values (1), (2), (3), (4);
+
+select * from ice01;
+
+-- create one branches
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/test/queries/positive/replace_iceberg_branch.q:11]:
```suggestion
-- create one branche
```

File: parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g (hunk 1 above)
Comment (ayushtkn) [reply parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:520]: I will add these. Was checking the spark syntax, they allow only replacing with snapshot id, not with branch?

Comment (ayushtkn) [reply parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:520]: So, I will change the syntax like:
```
ALTER TABLE prod.db.sample REPLACE BRANCH audit-branch
AS OF SYSTEM_VERSION 4567 RETAIN 60 DAYS
```
and for branch
```
ALTER TABLE prod.db.sample REPLACE BRANCH audit-branch
AS OF BRANCH branch-2 RETAIN 60 DAYS
```

I think we are not using version but system_version everywhere.

**Comment by ayushtkn** [comment]
> I think for hive table, we can reuse overwrite  to emulate replace syntax, and for iceberg table, we can just use iceberg api to do this. We can support iceberg table first.

Sounds like IOW with COW to me, that uses OverwriteFiles only I believe

File: parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g (hunk 1 above)
Comment (zhangbutao) [reply parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g:520]: https://github.com/apache/iceberg/pull/6638/files#diff-7bded55cc5d4893d2eb9d899ac095939e7b47814374d21cff159cb7ded9a503e  Check spark implementation, looks like` spark replace `has more properties, e.g. retain, retention..

File: iceberg/iceberg-handler/src/test/queries/positive/replace_iceberg_branch.q (hunk 2 above)
Comment (ayushtkn) [reply iceberg/iceberg-handler/src/test/queries/positive/replace_iceberg_branch.q]: done

File: parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g (hunk 1 above)
Comment (ayushtkn) [reply parser/src/java/org/apache/hadoop/hive/ql/parse/AlterClauseParser.g]: I have copied those from the create branch & pushed a new change, give a check once if the approach looks fine

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/snapshotref/branch/replace/AlterTableReplaceBranchRefAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+          replaceSnapshotrefSpec.setMaxSnapshotAgeMs(maxSnapshotAgeMs);
+        }
+        break;
+      default:
+        throw new SemanticException("Unrecognized token in ALTER " + alterTableType.getName() + " statement");
+      }
+    }
+
+    AlterTableSnapshotRefSpec<AlterTableSnapshotRefSpec.RenameSnapshotrefSpec> alterTableSnapshotRefSpec =
+        new AlterTableSnapshotRefSpec(alterTableType, replaceSnapshotrefSpec);
```
Comment (zhangbutao) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/snapshotref/branch/replace/AlterTableReplaceBranchRefAnalyzer.java:100]: `RenameSnapshotrefSpec`? Shouldn't this be the `ReplaceSnapshotrefSpec`?
```suggestion
    AlterTableSnapshotRefSpec<AlterTableSnapshotRefSpec.ReplaceSnapshotrefSpec> alterTableSnapshotRefSpec =
        new AlterTableSnapshotRefSpec<>(alterTableType, replaceSnapshotrefSpec);
```

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/snapshotref/branch/replace/AlterTableReplaceBranchRefAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  @Override
+  protected void analyzeCommand(TableName tableName, Map<String, String> partitionSpec, ASTNode command)
+      throws SemanticException {
+    Table table = getTable(tableName);
+    DDLUtils.validateTableIsIceberg(table);
+    inputs.add(new ReadEntity(table));
+    validateAlterTableType(table, alterTableType, false);
+    AlterTableSnapshotRefSpec.ReplaceSnapshotrefSpec replaceSnapshotrefSpec;
+    String sourceBranch = command.getChild(0).getText();
+    int i;
```
Comment (zhangbutao) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/snapshotref/branch/replace/AlterTableReplaceBranchRefAnalyzer.java:62]: Could we give a more meaningful name for this? e.g. `childNodeNum`? or any other valid naming.

**Review by zhangbutao — COMMENTED** [review]
@ayushtkn I am fine with this change. Left minor comments & pending green ci.
Thanks.

**Review by zhangbutao — APPROVED** [review]
LGTM +1

**Comment by zhangbutao** [comment]
> http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5190/3/tests I think `mv_iceberg_orc5.q` is a flaky test, as i have seen it in multiple CI. e.g. http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5195/4/tests
> 
> I just run a flaky test to check `mv_iceberg_orc5.q` : http://ci.hive.apache.org/job/hive-flaky-check/837/

In fact, `mv_iceberg_orc5.q`  can run successfully, but it has different execution planning compared with existing q.out. So maybe the flaky ci test can not check if `mv_iceberg_orc5.q` is flaky. :(

**Comment by zhangbutao** [comment]
> http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5190/3/tests I think `mv_iceberg_orc5.q` is a flaky test, as i have seen it in multiple CI. e.g. http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5195/4/tests
> 
> I just run a flaky test to check `mv_iceberg_orc5.q` : http://ci.hive.apache.org/job/hive-flaky-check/837/

http://ci.hive.apache.org/job/hive-flaky-check/837/testReport/ Test report shows that  `mv_iceberg_orc5.q` is a flaky test.

**Pull request merged**